### ISSUE TITLE
(Feature request)  Dynamic relations, or attribute inheritance between linked notes

### ISSUE THREAD

**Issue by Potjoe-97** [zadam/trilium#3493]
### Describe feature

Hi, it is really possible this feature request needs to go to the discussion section, please advise me if it's the case!

The idea is the following. In my daily use, the tree concept in trillium is a great way for _storing notes_ but less efficient for _linking notes_ (which I find relations attributes to do a better job of). However, my poor scripting capabilities, and the current architecture, doesn't allow to share attributes between two linked notes : attribute inheritance is only designed with the tree "parent-child" relation in mind, and  I think it would benefit to be open to other relations. 

### Additional Information

Let's have a real use case as an example. 
- In the subtree "Authors", we have notes for author A, B, etc
- In each author notes, book notes are created with a relation attribute "Author" pointing to their parent. 
- In another subtree, "Extract" notes, each have a relation attribute linking to their respective "Book" note"

Instead of having to move every extract note to their respective book note for them to inherit the "Author" relation (which is tedious when they are scattered), it would be possible to share the "Author" relation from the book to the "Extract" note.

**Comment by zadam** [comment]
Hi, would [template](https://github.com/zadam/trilium/wiki/Template) possibly work for you? It has some other features which you can ignore, but for your use case this is important:

> all attributes from the template note are inherited to the instance notes

This allows inheriting attributes not by tree structure, but by a "template" relation. If your "extract" note points with "template" relation to the "book" note, then all "book"'s attributes will be inherited to the "extract" note.

**Comment by zadam** [comment]
> I'd loose the possibility to set attributes for all extract notes at the same time, since each of them would point to a different template (the corresponding book), and I think it would make them more difficult to search for.

Unless I'm missing something, I don't think that's the case:

* You can use multiple template relations on a note, and the note will inherit attributes from all of them.
* template notes can themselves have a template relation, inheriting (also transitively) attributes. You can form essentially a template (type) inheritance tree. It's an overkill for most needs, though.

I've been thinking about this and believe that "template" relation usage might be complicated for this use case, because it's an amalgamation of features - inheriting attributes is here desirable, but its other properties - copying content / subtree might prove problematic. I've now added an "inherit" relation which separates this and provides only attribute inheritance.

> For instance, be able to define a relation such as ~Author=CurrentNote.relations.Extract.relations.Author.noteId in the Extract note template would do the trick. Is this something feasible and / or conceivable in the current state of the project ?

There are two possibilities here:

* the relation would be truly dynamic - resolved at runtime. In general, more gotchas and complexity, completely new relation type
* `CurrentNote.relations.Extract.relations.Author.noteId` would be a sort of "relation-generator" which would at some point of time produce a real relation based on this path. A good time point would be "template instantiation" - as part of that the template's subtree is duplicated into the instance and these generators could be "executed" and thus converted into real relations.

**Comment by Potjoe-97** [comment]
> * You can use multiple template relations on a note

Indeed, I did not know it was possible ! I take note of your warning regarding other features though (I did not noticed it copied the whole substree, since I'm only using them as "true" empty templates). 
Thank you for implementing the new "inherit" relation If I understand you correctly, it will allow to inherit attributes only from a note to another, will it ? 

> * `CurrentNote.relations.Extract.relations.Author.noteId` would be a sort of "relation-generator" which would at some point of time produce a real relation based on this path.

The concept of "relation-generator" seems great. Indeed it doesn't need to be refreshed at any time anyway, but only to be executed once to convert the argument into a relation with a real path to a note. 


> * A good time point would be "template instantiation"

Not totally sure about this, unless there is something I don't understand. I agree that this could be useful at the template level. However, after thinking about it : 
* If the generator is executed at template instantiation, it implies that every relations value used by the generator must be set _before_ the template is instantiated (and the generator executed) : it that case, it seems to be easier to set up the relation value in the template in the first place. 
* If the generator is executed any time _after_ template instantiation, these relation-generator would have a broader variety of usage. For instance, you could have a promoted attribute defined in the template, link a note to the template, enter the value of the promoted attribute, and use these relation generator to assign a relation value based on the promoted attribute value. 

Maybe we can imagine to have the trigger configurable : run at template instantiation, run at attribute change, etc..

**Comment by zadam** [comment]
> Thank you for implementing the new "inherit" relation If I understand you correctly, it will allow to inherit attributes only from a note to another, will it ?

Yes. It's just like "template" without copying note content and subtree. 

> it that case, it seems to be easier to set up the relation value in the template in the first place.

But then it's set to a fixed value. Imagine the relation is simply "parent" - without generator, you can set it to exactly 1 note (template's parent, presumably). With generator, you can set it to "currentNote.parent" which will resolve after you create an "instance" of this template somewhere in the tree - the generator will create a "parent" relation pointing to the parent note of the template instance (not to the parent note of the template).

This would be useful in "template subtrees" where the instance notes get a copy of the whole subtree with notes already interlinked between each other with relations, e.g.: template:

* Book (with relation generator to the child summary)
  * Summary  

> For instance, you could have a promoted attribute defined in the template, link a note to the template, enter the value of the promoted attribute, and use these relation generator to assign a relation value based on the promoted attribute value.

I see what you mean. Determining the correct point of time to run would be a problem, though, and having to configure this just so it works is a bad sign.

**Comment by Potjoe-97** [comment]
Hi! Sorry for the absence, I had a lot going on recently.
